mayastor/src/bdev/dev/nvmx/channel.rs

```diff
@@ -0,0 +1,581 @@
+            }
+            IoType::Unmap => {
+                self.io_stats.num_unmap_ops += num_ops;
+                self.io_stats.bytes_unmapped += num_blocks;
+            }
+            _ => {
+                warn!("Unsupported I/O type for I/O statistics: {:?}", op);
+            }
+        }
+    }
+
+    /// Get I/O statistics for channel.
+    #[inline]
+    pub fn get_io_stats(&self) -> BlockDeviceIoStats {
+        let mut stats = self.io_stats;
+
+        // Translate sectors into bytes before returning the stats.
+        stats.bytes_read *= self.block_size;
+        stats.bytes_written *= self.block_size;
+        stats.bytes_unmapped *= self.block_size;
+
+        stats
+    }
+}
+
+pub struct NvmeControllerIoChannel(NonNull<spdk_io_channel>);
+
+extern "C" fn disconnected_qpair_cb(
+    qpair: *mut spdk_nvme_qpair,
+    _ctx: *mut c_void,
+) {
+    warn!(?qpair, "NVMe qpair disconnected");
+    /*
+     * Currently, just try to reconnect indefinitely. If we are doing a
+     * reset, the reset will reconnect a qpair and we will stop getting a
+     * callback for this one.
+     */
+    unsafe {
+        spdk_nvme_ctrlr_reconnect_io_qpair(qpair);
+    }
+}
+
+extern "C" fn nvme_poll(ctx: *mut c_void) -> i32 {
+    let inner = NvmeIoChannel::from_raw(ctx).inner_mut();
+
+    let num_completions = unsafe {
+        spdk_nvme_poll_group_process_completions(
+            inner.poll_group.as_ptr(),
+            0,
+            Some(disconnected_qpair_cb),
+        )
+    };
+
+    if num_completions > 0 {
+        1
+    } else {
+        0
+    }
+}
+
+impl NvmeControllerIoChannel {
+    pub extern "C" fn create(device: *mut c_void, ctx: *mut c_void) -> i32 {
+        let id = device as u64;
+
+        debug!("Creating IO channel for controller ID 0x{:X}", id);
+
+        let carc = match NVME_CONTROLLERS.lookup_by_name(id.to_string()) {
+            None => {
+                error!("No NVMe controller found for ID 0x{:X}", id);
+                return 1;
+            }
+            Some(c) => c,
+        };
+
+        let (cname, spdk_handle, block_size) = {
+            let controller = carc.lock().expect("lock error");
+            // Make sure controller is available.
+            if controller.get_state() != NvmeControllerState::Running {
+                error!(
+                    "{} controller is in {:?} state, I/O channel creation not possible",
+                    controller.get_name(),
+                    controller.get_state()
+                );
+                return 1;
+            }
+            // Release controller's lock before proceeding to avoid deadlocks,
+            // as qpair-related operations might hang in case of
+            // network connection failures. Note that we still hold
+            // the reference to the controller instance (carc) which
+            // guarantees that the controller exists during I/O channel
+            // creation.
```

Comment:
having the arc guarantees the the controller exists but not that it hasn't yet been removed?

Reply:
Yes, maybe it is worth using more or less the same naming scheme as in libspdk:
```
return rc == 0 ? SPDK_POLLER_IDLE : SPDK_POLLER_BUSY;
```